### PR TITLE
Added custom strategies possibility

### PR DESCRIPTION
**Pull request by redwarewolf**
- Added possibility to create custom strategies and custom base case
- Added new tests
- Added Interface module
- Added a Strategy Interface for custom strategies

### CODE REVIEW

**Comment by holywyvern** [comment]
I know than using a design by contract is useful, people suggesdted better patterns for them, you may consider changes for a 2.0 release if they bring improvements.

File: lib/wor/batchifier.rb [hunk 1]
```diff
@@ -1,14 +1,20 @@
...
+
 module Wor
   module Batchifier
-    class WrongStrategy < StandardError; end
 
     # This module exports the function execute_in_batches, that needs a collections and
     # => optionaly a batch_size and a merge strategy. It will slice the collection and
-    # => apply the given block to all chunks and merge the results. It expects the responses
-    # => to be Hash. It can ignore them if the given stragegy is no_response
+    # => apply the chozen strategy to all chunks and merge the results. It expects the responses
```
Comment (leoncalermo) [inline lib/wor/batchifier.rb]: I think you wanted to say `chosen` ? haha

File: lib/wor/batchifier.rb [hunk 2]
```diff
@@ -1,14 +1,20 @@
...
+    # => to be Hash. It can ignore them if the given strategy is no_response
     def execute_in_batches(collection, batch_size: 100, strategy: :add)
-      collection.each_slice(batch_size).to_a.inject({}) do |rec, chunk|
+      strategy_class = classify_strategy(strategy)
+      collection.each_slice(batch_size).to_a.inject(strategy_class.new.base_case) do |rec, chunk|
```
Comment (leoncalermo) [inline lib/wor/batchifier.rb]: i dont get why you turn it into an array, when you should be able to do it  as an enumerator

File: lib/wor/batchifier.rb [hunk 3]
```diff
@@ -17,10 +23,14 @@ def execute_in_batches(collection, batch_size: 100, strategy: :add)
...
     def merge(response, rec, strategy)
-      return rec.merge(response) { |_, v1, v2| v1 + v2 } if strategy == :add
-      return rec.merge(response) { |_, v1, _| v1 } if strategy == :maintain_unique
-      return {} if strategy == :no_response
-      raise WrongStrategy
+      return Wor::Batchifier.classify_strategy(strategy).new.merge_strategy(response,rec)
+    end
+
+    def classify_strategy(strategy)
+      strategy_class_name = strategy.to_s.split('_').collect(&:capitalize).join
```
Comment (leoncalermo) [inline lib/wor/batchifier.rb]: i think that `stategy.to_s.classify` does the same here

File: lib/wor/batchifier.rb (hunk 2 above)
Comment (redwarewolf) [reply lib/wor/batchifier.rb]: Yeah, this was fixed in the lazy_array PR, I dont use the to_a anymore and just work with enumerators.

File: lib/wor/batchifier.rb (hunk 3 above)
Comment (redwarewolf) [reply lib/wor/batchifier.rb]: It's pure ruby so it doesn't have the method classify which is from rails

File: lib/wor/batchifier.rb (hunk 1 above)
Comment (redwarewolf) [reply lib/wor/batchifier.rb]: Yes :see_no_evil:

**Pull request merged**